Sources/XiEditor/LoggingUtilities.swift

```diff
@@ -0,0 +1,79 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+
+/// Error tolerant wrapper for append-writing to a file.
+struct FileWriter {
```

Comment:
`FileWriter` consists of `path: URL` which is a value type, and `handle: FileHandle` which is a reference type.
Is there a real reason why it is a `struct`?

Reply:
No good reason, no. My default habit is to use a struct anywhere I won't ever want to pass a reference around. Is having a reference type in a struct not idiomatic?

Reply:
I think the main point is that `structs` have value semantics, which means copy-on-write. Copying a struct with underlaying reference type means that we have to handle the copy of the underlying reference type.

```swift
struct S {
    let ref: FileWrapper
    
    init() {
        ref = FileWrapper()
    }
}

let s = S()
print(s.ref) //"<NSFileWrapper: 0x6000011fd700>\n"

let copy = s

print(copy.ref) //"<NSFileWrapper: 0x6000011fd700>\n"
```

Notice that it's are sharing the reference, which breaks the expected behaviour of a `struct`.

Reply:
Okay, thanks!